code/modules/integrated_electronics/core/assemblies.dm

```diff
@@ -0,0 +1,372 @@
+	if(!check_interactivity(user))
+		return
+
+	var/total_parts = 0
+	var/total_complexity = 0
+	for(var/obj/item/integrated_circuit/part in contents)
+		total_parts += part.size
+		total_complexity = total_complexity + part.complexity
+	var/HTML = list()
+
+	HTML += "<html><head><title>[src.name]</title></head><body>"
+	HTML += "<br><a href='?src=[REF(src)]'>\[Refresh\]</a>  |  "
+	HTML += "<a href='?src=[REF(src)];rename=1'>\[Rename\]</a><br>"
+	HTML += "[total_parts]/[max_components] ([round((total_parts / max_components) * 100, 0.1)]%) space taken up in the assembly.<br>"
+	HTML += "[total_complexity]/[max_complexity] ([round((total_complexity / max_complexity) * 100, 0.1)]%) maximum complexity.<br>"
+	if(battery)
+		HTML += "[round(battery.charge, 0.1)]/[battery.maxcharge] ([round(battery.percent(), 0.1)]%) cell charge. <a href='?src=[REF(src)];remove_cell=1'>\[Remove\]</a>"
+	else
+		HTML += "<span class='danger'>No powercell detected!</span>"
+	HTML += "<br><br>"
+	HTML += "Components:<hr>"
+	HTML += "Built in:<br>"
+
+
+//Put removable circuits in separate categories from non-removable
+	for(var/obj/item/integrated_circuit/circuit in contents)
+		if(!circuit.removable)
+			HTML += "<a href=?src=[REF(circuit)];examine=1;from_assembly=1>[circuit.displayed_name]</a> | "
+			HTML += "<a href=?src=[REF(circuit)];rename=1;from_assembly=1>\[Rename\]</a> | "
+			HTML += "<a href=?src=[REF(circuit)];scan=1;from_assembly=1>\[Scan with Debugger\]</a> | "
+			HTML += "<a href=?src=[REF(circuit)];bottom=[REF(circuit)];from_assembly=1>\[Move to Bottom\]</a>"
+			HTML += "<br>"
+
+	HTML += "<hr>"
+	HTML += "Removable:<br>"
+
+	for(var/obj/item/integrated_circuit/circuit in contents)
+		if(circuit.removable)
+			HTML += "<a href=?src=[REF(circuit)];examine=1;from_assembly=1>[circuit.displayed_name]</a> | "
+			HTML += "<a href=?src=[REF(circuit)];rename=1;from_assembly=1>\[Rename\]</a> | "
+			HTML += "<a href=?src=[REF(circuit)];scan=1;from_assembly=1>\[Scan with Debugger\]</a> | "
+			HTML += "<a href=?src=[REF(circuit)];remove=1;from_assembly=1>\[Remove\]</a> | "
+			HTML += "<a href=?src=[REF(circuit)];bottom=[REF(circuit)];from_assembly=1>\[Move to Bottom\]</a>"
+			HTML += "<br>"
+
+	HTML += "</body></html>"
+	user << browse(jointext(HTML,null), "window=assembly-\[REF(src)];size=600x350;border=1;can_resize=1;can_close=1;can_minimize=1")
+
+/obj/item/device/electronic_assembly/Topic(href, href_list[])
+	if(..())
+		return 1
+
+	if(href_list["rename"])
+		rename(usr)
+
+	if(href_list["remove_cell"])
+		if(!battery)
+			to_chat(usr, "<span class='warning'>There's no power cell to remove from \the [src].</span>")
+		else
+			var/turf/T = get_turf(src)
+			battery.forceMove(T)
+			playsound(T, 'sound/items/Crowbar.ogg', 50, 1)
+			to_chat(usr, "<span class='notice'>You pull \the [battery] out of \the [src]'s power supplier.</span>")
+			battery = null
+
+	interact(usr) // To refresh the UI.
+
+/obj/item/device/electronic_assembly/verb/rename()
```

Comment:
VERRRRRRBSS

make this a proc called when hit by a pen instead

Reply:
Renames are performed with internal interaction in assembly.Not by pen.

Reply:
This still shouldn't be a verb, it should be a proc